packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/injecter.ts

```diff
@@ -0,0 +1,154 @@
+  grant(taskDefinition: ecs.TaskDefinition): void;
+}
+
+/**
+ * The settings for the `InjectableTopic` class.
+ */
+export interface InjectableTopicProps {
+  /**
+   * The SNS Topic to publish events to.
+   */
+  readonly topic: sns.ITopic;
+}
+
+/**
+ * The `InjectableTopic` class represents SNS Topic resource that can be published events to by the parent service.
+ */
+
+export class InjectableTopic implements GrantInjectable {
+  public readonly topic: sns.ITopic;
+
+  constructor(props: InjectableTopicProps) {
+    this.topic = props.topic;
+  }
+
+  public grant(taskDefinition: ecs.TaskDefinition) {
+    this.topic.grantPublish(taskDefinition.taskRole);
+  }
+
+  public environmentVariables(): { [key: string]: string } {
+    let environment: { [key: string]: string } = {};
+    environment[`${this.topic.node.id.toUpperCase()}_TOPIC_ARN`] = this.topic.topicArn;
+    return environment;
+  }
+}
+
+/**
+ * The settings for the Injecter extension.
+ */
+export interface InjecterExtensionProps {
+  /**
+   * The list of injectable resources for this service.
+   */
+  readonly injectables: Injectable[];
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to add the injectable resource environment variables.
+ */
+interface ContainerMutatingProps {
+  /**
+   * The resource environment variables to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the injectable resource environment variables.
+ */
+class InjecterExtensionMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension accepts a list of `Injectable` resources that the parent service can publish events or write data to.
+ * It sets up the corresponding permissions for the task role of the parent service.
+ */
+export class InjecterExtension extends ServiceExtension {
+  private props: InjecterExtensionProps;
+
+  private environment: { [key: string]: string } = {};
+
+  constructor(props: InjecterExtensionProps) {
+    super('injecter');
+
+    this.props = props;
+  }
+
+  // @ts-ignore - Ignore unused params that are required for abstract class extend
+  public prehook(service: Service, scope: Construct) {
+    this.parentService = service;
+
+    for (const injectable of this.props.injectables) {
+      for (const [key, val] of Object.entries(injectable.environmentVariables())) {
+        this.environment[`${service.id.toUpperCase()}_${key}`] = val;
```

Comment:
so will this create an environment variable like:
```
PUBLISHER_MYTOPIC_TOPIC_ARN
```

Was there a reason why you prefered to prefix with the `service.id`? I think customers that implement the `Injectable` interface might expect their environment variable to appear as is 💭

Reply:
Yes, it creates `PUBLISHER_MYTOPIC_TOPIC_ARN` . 

The `service.id` was added as a prefix to distinguish between topics in cases of collision, when two topics with same name are created in different services.

Reply:
gotcha, I think since this works a bit different than copi where users create their own topics, I think we should pass the env var as is!

Reply:
Makes sense!